src/influxdb_ioxd/rpc/storage/data.rs

```diff
@@ -275,52 +283,80 @@ fn convert_tags(table_name: &str, field_name: &str, tags: &[(Arc<str>, Arc<str>)
 }
 
 trait ExtractValues<T> {
-    /// Extracts num_rows of data starting from start_row as a vector
-    fn extract_values(&self, start_row: usize, num_rows: usize) -> Vec<T>;
+    /// Extracts num_rows of data starting from start_row as a vector,
+    /// for all rows `i` where `valid[i]` is set
+    fn extract_values(&self, start_row: usize, num_rows: usize, valid: Option<&Bitmap>) -> Vec<T>;
+}
+
+/// Implements extract_values for a particular type of array that
+macro_rules! impl_extract_values {
+    ($DATA_TYPE:ident) => {
+        fn extract_values(
+            &self,
+            start_row: usize,
+            num_rows: usize,
+            valid: Option<&Bitmap>,
+        ) -> Vec<$DATA_TYPE> {
+            let end_row = start_row + num_rows;
+            match valid {
+                Some(valid) => (start_row..end_row)
+                    .filter_map(|row| {
+                        if valid.is_set(row) {
+                            Some(self.value(row))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect(),
+                None => (start_row..end_row).map(|row| self.value(row)).collect(),
+            }
+        }
+    };
 }
 
 impl ExtractValues<String> for StringArray {
-    fn extract_values(&self, start_row: usize, num_rows: usize) -> Vec<String> {
+    fn extract_values(
+        &self,
+        start_row: usize,
+        num_rows: usize,
+        valid: Option<&Bitmap>,
+    ) -> Vec<String> {
         let end_row = start_row + num_rows;
-        (start_row..end_row)
-            .map(|row| self.value(row).to_string())
-            .collect()
+        match valid {
+            Some(valid) => (start_row..end_row)
+                .filter_map(|row| {
+                    if valid.is_set(row) {
+                        Some(self.value(row).to_string())
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+            None => (start_row..end_row)
+                .map(|row| self.value(row).to_string())
+                .collect(),
```

Comment:
Another place for [then](https://doc.rust-lang.org/std/primitive.bool.html#method.then):

```suggestion
            Some(valid) => (start_row..end_row)
                .filter_map(|row| valid.is_set(row).then(|| self.value(row).to_string()))
                .collect(),
            None => (start_row..end_row)
                .map(|row| self.value(row).to_string())
                .collect(),
```